### PR TITLE
Add Joint Training

### PR DESCRIPTION
**Pull request by anwai98**

### CODE REVIEW

File: micro_sam/training/joint_sam_trainer.py [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+
+class JointSamTrainer(SamTrainer):
+    def __init__(
+            self, **kwargs
+    ):
+        super().__init__(**kwargs)
+        self.unetr = UNETRForJointTraining(
+            img_size=self.model.img_size,
+            backbone="sam",
+            encoder="vit_b",
```
Comment (constantinpape) [inline micro_sam/training/joint_sam_trainer.py:81]: I think we need to extract the encoder from the sam model and then pass it here.

File: micro_sam/training/joint_sam_trainer.py [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+
+            with forward_context():
+                # 1. train for the interactive segmentation
+                (loss, mask_loss, iou_regression_loss, model_iou,
+                 sampled_binary_y) = self._interactive_train_iteration(x, y, self._iteration)
+
+            backprop(loss)
+
+            # let's get the unetr decoder for doing the instance segmentation
+            self.unetr.encoder = self.model.encoder  # TODO: revisit
```
Comment (constantinpape) [inline micro_sam/training/joint_sam_trainer.py:114]: This should not be necessary if you initialize the unetr correctly.

Comment (anwai98) [reply micro_sam/training/joint_sam_trainer.py]: I still need to have a closer look at this. Will keep you posted.

File: micro_sam/training/joint_sam_trainer.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+                    # 1. validate for the interactive segmentation
+                    (loss, mask_loss, iou_regression_loss, model_iou,
+                     sampled_binary_y, metric) = self._interactive_val_iteration(x, labels_instances, val_iteration)
+
+                with forward_context():
+                    # 2. validate for the automatic instance segmentation
+                    unetr_loss = self._instance_train_iteration(x, labels_for_unetr)
+
+                loss_val += loss.item()
+                metric_val += metric.item()
```
Comment (constantinpape) [inline micro_sam/training/joint_sam_trainer.py:102]: Potentially we could also add the unetr loss as metric here. (Let's discuss tomorrow.)

**Review by constantinpape — COMMENTED** [review]
Looks good! Implementation of the optimizer is exactly what I meant.

File: micro_sam/training/joint_sam_trainer.py (hunk 3 above)
Comment (anwai98) [reply micro_sam/training/joint_sam_trainer.py:102]: Yes, that's something I wanted to discuss. Let's check in tomorrow. Thanks!

Comment (anwai98) [reply micro_sam/training/joint_sam_trainer.py]: This is taken care of now. Thanks!

**Comment by anwai98** [comment]
This PR now looks good to go as well. This includes:
- Refactoring the `SamTrainer`
- Adding `JointSamTrainer` for the joint training setup
- Adding a script for training within the joint training framework.
- (Misc.) Performing automatic instance segmentation using the UNETR weights from finetuned specialist models.

**Review by constantinpape — APPROVED** [review]
